src/IO/DataImporter/DataFileReader.hpp

```diff
@@ -0,0 +1,76 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include "AlgorithmNodegroup.hpp"
+#include "IO/DataImporter/Tags.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
+#include "Parallel/ParallelComponentHelpers.hpp"
+#include "Parallel/PhaseDependentActionList.hpp"
+#include "ParallelAlgorithms/Initialization/MergeIntoDataBox.hpp"
+
+namespace importer {
```

Comment:
It's initial data for evolutions and an initial guess for elliptic solves.. the component is quite general; it can read a volume data file and populate tensors in each element's DataBox from it. So how about `DataImporter` (that's currently the directory name), or perhaps you have other suggestions?

Reply:
I don't love `DataImporter` since that's literally anything that reads things from anywhere. How is the initial guess not initial data for an elliptic solve?

Reply:
Could be misleading to do an (elliptic) initial data solve and import „initial data“ for the initial guess, don‘t you think?

What’s the inverse to `Observers`? How about `TensorDataImporter` or `VolumeDataImporter`? A shorter name would be better..
Or we place it in the `observers` namespace since it’s observers in reverse?

Reply:
I don't think it's confusing. I think the issue people are confused about is that people think elliptic problems are only for initial data :roll_eyes: 

It should definitely not be in observers namespace since it's not writing data to disk.

Reply:
So we have two options that may be mildly confusing in different ways:
- `importer` doesn’t make clear it’s for volume tensor data
- `initial_data_importer` may be confusing when solving for initial data
I’m actually not quite sold on `initial_data_importer`.. it’s very long, we don’t have any other namespace that’s this long or has underscores, instead we commonly use abbreviations for namespaces. Also, the importer is not really limited to initial data; it can load tensor data at any time.
How about `importer`, `import`, `importing` or `imprt`, and I name the parallel component `VolumeDataFileReader` instead? The we can use the namespace for other possible data importing components that aren’t tensor volume data, and in `IO` we have a namespace for “in” and one for “out” :)

Reply:
I don't like `initial_data_importer` because there is nothing that ties this to initial data.  It literally just reads in data the has been written in the `h5::VolumeData` format.  Therefore in theory it could be used as the reader for something like SpEC's ApplyObservers.  

The new namespace is to be the reading equivalent of `observers`
The new component is to be the reading equivalent of `observers::ObserverWriter`, but specifically for volume data.
The new action is to be the reading equivalent of `observers::ThreadedActions::WriteVolumeData`

I would suggest WriteVolumeData -> ReadVolumeData,  ObserverWriter -> VolumeDataReader, and I can't think of a better name than `importers` for the namespace...

Reply:
If my understanding of the code is correct it does not just read in data, it assumes that there is an element array component. I suspect we will find it rather limited in its utility

Reply:
Here's why I don't think this code should be viewed as anything more general than reading initial data:

1. As far as I can tell it requires the exact element structure to present in the file and so does not do any interpolation.
2. It requires creating lots of elements in a parallel component in order to read the data to begin with.
3. For ApplyObservers style transformations we almost certainly don't want to create each element and split the data, we would want to operate on the collection of data in the file. If the data is too large for a single node we would want a more efficient splitting than each element.
4. There's no infrastructure for creating elements based on the contents of the file as far as I can tell. This restricts the code even in terms of reading in initial data.

Reply:
Yes, the current implementation simply “observes in reverse”. We’ll probably add features such as interpolations once we need them. So how about we do such things in an `importers` namespace, like we have components for reductions and volume data in `observers`?

Reply:
Sure, I'd like to sketch out the extra features we want to have and then file an issue as part of this PR. Would that be okay?

What I think we want (in no particular order):

- Support for interpolating to an arbitrary grid.
- Support for extrapolation to grid points if desired (BBH initial data used to do this in SpEC, I'm not sure it still does)
- The option to retrieve the data across all elements of a particular tensor or tensor component stored in a particular file. (I'm not sure how much this can be done in the same framework, might just need an extra action).
- Create elements of a DG element array from the elements in the file(s). This will be useful for getting an initial AMR hierarchy for the evolution code from the elliptic solver AMR grid used while solving for initial data.

Anything else?

Reply:
Sounds good to me. #1926

Reply:
Sounds good to me. #1926